### PR TITLE
Backport #13202 to branch_9x

### PR DESCRIPTION
**Pull request by vigyasharma**
- Add timeout support to AbstractKnnVectorQuery (#13202)
- Fix failing BaseKnnVectorQueryTestCase#testTimeout (#13283)

### CODE REVIEW

**Comment by vigyasharma** [comment]
Now that the test failure have been fixed, I'm backporting #13202 to `branck_9x`. Running into some build issues on my local setup, which I think are unrelated to the commit. Opening this PR to run precommit checks before backporting.

**Comment by vigyasharma** [comment]
Getting the same test failure here.

```java
> Task :lucene:core:compileTestJava
/home/runner/work/lucene/lucene/lucene/core/src/test/org/apache/lucene/search/BaseKnnVectorQueryTestCase.java:816: error: cannot find symbol
          noTimeoutManager.newCollector(Integer.MAX_VALUE, searcher.leafContexts.getFirst());
                                                                                ^
  symbol:   method getFirst()
  location: variable leafContexts of type List<LeafReaderContext>
/home/runner/work/lucene/lucene/lucene/core/src/test/org/apache/lucene/search/BaseKnnVectorQueryTestCase.java:832: error: cannot find symbol
          timeoutManager.newCollector(Integer.MAX_VALUE, searcher.leafContexts.getFirst());
                                                                              ^
  symbol:   method getFirst()
  location: variable leafContexts of type List<LeafReaderContext>
Note: Some input files use or override a deprecated API.
```

File: lucene/core/src/test/org/apache/lucene/search/BaseKnnVectorQueryTestCase.java [hunk 1]
```diff
@@ -797,6 +800,77 @@ public void testBitSetQuery() throws IOException {
...
+        IndexReader reader = DirectoryReader.open(indexStore)) {
+      IndexSearcher searcher = newSearcher(reader);
+
+      KnnCollectorManager delegate = new TopKnnCollectorManager(3, searcher);
+
+      // A collector manager with no timeout
+      TimeLimitingKnnCollectorManager noTimeoutManager =
+          new TimeLimitingKnnCollectorManager(delegate, null);
+      KnnCollector noTimeoutCollector =
+          noTimeoutManager.newCollector(Integer.MAX_VALUE, searcher.leafContexts.getFirst());
```
Comment (benwtrent) [inline lucene/core/src/test/org/apache/lucene/search/BaseKnnVectorQueryTestCase.java:816]: Pretty sure `getFirst` is a JDK21+ only thing. The compilation level on branch_9x is too low to allow that API.

**Comment by vigyasharma** [comment]
Backported changes directly to retain original commit messages. This PR would've needed a squash merge as it was not rebased on the latest branch_9x commit.

**Pull request closed without merging**